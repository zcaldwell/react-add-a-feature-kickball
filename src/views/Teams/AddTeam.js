import { useState } from 'react';
import { Link, useHistory } from 'react-router-dom';
import TeamForm from '../../components/teams/TeamForm';
import { createTeam } from '../../services/teams';

function AddTeam() {
  const [name, setName] = useState('');
  const [city, setCity] = useState('');
  const [state, setState] = useState('');
  const [alert, setAlert] = useState('');
  const history = useHistory();

  const onChange = ({ target }) => {
    switch (target.name) {
      case 'name':
        return setName(target.value);
      case 'city':
        return setCity(target.value);
      case 'state':
        return setState(target.value);
      default:
        return false;
    }
  };

  const onSubmit = async (event) => {
    event.preventDefault();

    try {
      const resp = await createTeam({ name, city, state });
      history.push(`/teams/${resp[0].id}`);
    } catch (error) {
      setAlert(error.message);
    }
  };

  return (
    <>
      <Link to="/teams">Teams</Link> &raquo; <span>Add New</span>
      <br />
      <br />
      <p>{alert}</p>
      <TeamForm
        label="Add Team"
        name={name}
        city={city}
        state={state}
        onChange={onChange}
        onSubmit={onSubmit}
      />
    </>
  );
}

export default AddTeam;
