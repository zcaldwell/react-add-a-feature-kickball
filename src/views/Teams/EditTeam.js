import { useEffect, useState } from 'react';
import { Link, useHistory, useParams } from 'react-router-dom';
import TeamForm from '../../components/teams/TeamForm';
import { getTeamById, updateTeamById } from '../../services/teams';

function EditTeam() {
  const [input, setInput] = useState({ name: '', city: '', state: '' });
  const [team, setTeam] = useState({ name: '', city: '', state: '' });
  const [alert, setAlert] = useState('');
  const history = useHistory();
  const { id } = useParams();

  useEffect(() => {
    setInput(team);
  }, [team]);

  useEffect(() => {
    const fetchData = async () => {
      const resp = await getTeamById(id);
      setTeam(resp);
    };
    fetchData();
  }, [id]);

  const onChange = ({ target }) => {
    setInput((prevState) => ({ ...prevState, [target.name]: target.value }));
  };

  const onSubmit = async (event) => {
    event.preventDefault();

    try {
      const resp = await updateTeamById(team.id, {
        name: input.name,
        city: input.city,
        state: input.state,
      });
      history.push(`/teams/${resp[0].id}`);
    } catch (e) {
      setAlert(e.message ? e.message : 'Something went wrong :(');
    }
  };

  return (
    <>
      <Link to="/teams">Teams</Link> &raquo; <Link to={`/teams/${id}`}>{team.name}</Link> &raquo;{' '}
      <span>Edit</span>
      <br />
      <br />
      <p>{alert}</p>
      <TeamForm
        label="Edit Team"
        name={input.name}
        city={input.city}
        state={input.state}
        onChange={onChange}
        onSubmit={onSubmit}
      />
    </>
  );
}

export default EditTeam;
