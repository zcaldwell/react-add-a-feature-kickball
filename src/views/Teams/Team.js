import { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import Loading from '../../components/loading/Loading';
import { getTeamById } from '../../services/teams';

function Team({
  match: {
    params: { id },
  },
}) {
  const [team, setTeam] = useState({});
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const fetchData = async () => {
      const resp = await getTeamById(id);
      setTeam(resp);
      setLoading(false);
    };
    fetchData();
  }, [id]);

  return loading ? (
    <Loading />
  ) : (
    <>
      <Link to="/teams">Teams</Link> &raquo; <span>{team.name}</span>
      <h1>{team.name}</h1>
      <p>
        {team.city}, {team.state}
      </p>
      <p>
        <Link to={`/teams/${id}/edit`}>Edit Team</Link>
      </p>
    </>
  );
}

export default Team;
