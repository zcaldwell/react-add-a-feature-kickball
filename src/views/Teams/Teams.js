import { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import Loading from '../../components/loading/Loading';
import { deleteTeamById, getTeams } from '../../services/teams';

function Teams() {
  const [teams, setTeams] = useState([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const fetchData = async () => {
      const resp = await getTeams();
      setTeams(resp);
      setLoading(false);
    };
    fetchData();
  }, []);

  const handleDelete = async ({ id, name }) => {
    const shouldDelete = confirm(`Are you sure you want to delete ${name}?`);

    if (shouldDelete) {
      await deleteTeamById(id);
      const resp = await getTeams();
      setTeams(resp);
    }
  };

  return (
    <>
      <Link to="/teams/new">Add New Team</Link>
      {loading ? (
        <Loading />
      ) : (
        <>
          <h1>Teams</h1>
          <table className="resource-table">
            <thead>
              <tr>
                <th>Team Name</th>
                <th>Actions</th>
              </tr>
            </thead>
            <tbody>
              {teams.map((team) => (
                <tr key={team.id}>
                  <td>
                    <Link to={`/teams/${team.id}`}>{team.name}</Link>
                  </td>
                  <td>
                    <Link to={`/teams/${team.id}`}>
                      <button type="button" className="btn-view">
                        View
                      </button>
                    </Link>

                    <Link to={`/teams/${team.id}/edit`}>
                      <button type="button" className="btn-edit">
                        Edit
                      </button>
                    </Link>
                    <button type="button" className="btn-delete" onClick={() => handleDelete(team)}>
                      Delete
                    </button>
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </>
      )}
    </>
  );
}

export default Teams;
