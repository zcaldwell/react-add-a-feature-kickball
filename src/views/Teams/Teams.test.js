import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Teams from './Teams';

it('should render a list of teams', async () => {
  // MemoryRouter is needed because <Teams> uses the <Link> component
  const { container } = render(
    <MemoryRouter>
      <Teams />
    </MemoryRouter>
  );
  screen.getByLabelText(/loading/);
  await screen.findByText(/Bridge City Sneakers/);

  expect(container).toMatchSnapshot();
});
