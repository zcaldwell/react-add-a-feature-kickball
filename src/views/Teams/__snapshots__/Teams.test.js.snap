// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a list of teams 1`] = `
<div>
  <a
    href="/teams/new"
  >
    Add New Team
  </a>
  <h1>
    Teams
  </h1>
  <table
    class="resource-table"
  >
    <thead>
      <tr>
        <th>
          Team Name
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <a
            href="/teams/2"
          >
            Bridge City Sneakers
          </a>
        </td>
        <td>
          <a
            href="/teams/2"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
          <a
            href="/teams/2/edit"
          >
            <button
              class="btn-edit"
              type="button"
            >
              Edit
            </button>
          </a>
          <button
            class="btn-delete"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/teams/7"
          >
            Goodlanders
          </a>
        </td>
        <td>
          <a
            href="/teams/7"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
          <a
            href="/teams/7/edit"
          >
            <button
              class="btn-edit"
              type="button"
            >
              Edit
            </button>
          </a>
          <button
            class="btn-delete"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/teams/3"
          >
            Lakeville Thunderfeet
          </a>
        </td>
        <td>
          <a
            href="/teams/3"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
          <a
            href="/teams/3/edit"
          >
            <button
              class="btn-edit"
              type="button"
            >
              Edit
            </button>
          </a>
          <button
            class="btn-delete"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/teams/4"
          >
            Mt. Hoodies
          </a>
        </td>
        <td>
          <a
            href="/teams/4"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
          <a
            href="/teams/4/edit"
          >
            <button
              class="btn-edit"
              type="button"
            >
              Edit
            </button>
          </a>
          <button
            class="btn-delete"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/teams/1"
          >
            Rip City Kickers
          </a>
        </td>
        <td>
          <a
            href="/teams/1"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
          <a
            href="/teams/1/edit"
          >
            <button
              class="btn-edit"
              type="button"
            >
              Edit
            </button>
          </a>
          <button
            class="btn-delete"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/teams/5"
          >
            Stumptown Lumberjacks
          </a>
        </td>
        <td>
          <a
            href="/teams/5"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
          <a
            href="/teams/5/edit"
          >
            <button
              class="btn-edit"
              type="button"
            >
              Edit
            </button>
          </a>
          <button
            class="btn-delete"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/teams/6"
          >
            Vancougars
          </a>
        </td>
        <td>
          <a
            href="/teams/6"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
          <a
            href="/teams/6/edit"
          >
            <button
              class="btn-edit"
              type="button"
            >
              Edit
            </button>
          </a>
          <button
            class="btn-delete"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
